### PR TITLE
Query Loop Block: Use BlockPatternsList component for Query block patterns modal

### LINKED ISSUE #33202: Query Loop block: improve setup state
I'd like to propose a couple of small improvements to the current placeholder of the block.

The first one is adding the title of the variation in the carousel view. While I don't think it's a critical improvement, I think it's important that the two views (carousel and grid) are consistent. 

I think there are two possible positions. The left one is probably more useful because it allows users to connect the title with the variation right away. 

image

My other proposal would be to improve the "Star blank" placeholder with two things:

- A backlink (so users can go back to the previous state of the placeholder)
- A form to configure the contents of each post.

image

The form is less confusing than the list of icons with labels (it doesn't require you to scan the full list to see if the combination you want exists) and gives a little bit of extra control without complicating the interface too much.

### PR DESCRIPTION
**Pull request by apeatling**
Fixes https://github.com/WordPress/gutenberg/issues/33202

## What?

Replace the `BlockPatternSetup` component with `BlockPatternsList` component to match the experience of selecting patterns in different locations across the editor. 

## Why?

Lots of feedback on the linked issue above. In a nutshell the current carousel style pattern selection feels broken and confusing to users because it is different from anything else.

By moving to `BlockPatternsList` we start the process of consolidating around that component and moving closer to a [consistent experience for pattern selection](https://github.com/WordPress/gutenberg/issues/39308#issuecomment-1133086932).

## How?

This PR separates out the pattern selection modal functionality and replaces it to use the `BlockPatternsList` component. It also moves the search functionality for filtering pattern data out of the template-part block and into the utils directory where it can be reused.

## Testing Instructions

1. Open the site editor and edit a template with a query loop (or insert a query loop).
2. Use the "Replace" button in the query loop block toolbar
3. Confirm you see a modal with a searchable list of patterns that can be used in the query loop block.
4. Use the search and confirm the list is filtered correctly.
5. Select a pattern and confirm that it correctly replaces the content of the query post block.
6. Confirm that unit tests pass.

## Screenshots or screencast 

| Before  | After |
| ------------- | ------------- |
| <video src="https://user-images.githubusercontent.com/1464705/214386086-c1a56860-288e-49fd-bb4d-681473a5fc2a.mp4">  | <video src="https://user-images.githubusercontent.com/1464705/214386098-69bb5543-8f6c-49f6-b596-0b35d735fad1.mp4">  |

### CODE REVIEW

**Comment by apeatling** [comment]
I think this one is in a good spot now and ready for reviews :+1:

**Review by jameskoster — APPROVED** [review]
Big +1 to aligning these flows then refining holistically later.

This is working as advertised so I'm approving from the design side.

File: packages/block-library/src/query/edit/index.js
Comment (ntsekouras) [inline packages/block-library/src/query/edit/index.js:71]: We need to preserve this behavior. What this does in a few words is that shows the suggested patterns using the current `postType` set. For example if we have `postType: products` the insert/replace pattern flows show patterns with `products`.

Comment (apeatling) [reply packages/block-library/src/query/edit/index.js:71]: Thanks, good catch. I've added this back. Is there an easy way to test this? Maybe using WooCommerce?

Comment (ntsekouras) [reply packages/block-library/src/query/edit/index.js:71]: You can add the below in `query/variations.js` and the search for `Pages variation` and then click `choose`.

```
{
	name: 'pages-list',
	title: __( 'Pages variation' ),
	description: __(
		'Display a list of your most recent...'
	),
	icon: postList,
	attributes: {
		query: {
			perPage: 4,
			pages: 1,
			offset: 0,
			postType: 'page',
			order: 'desc',
			orderBy: 'date',
			author: '',
			search: '',
			sticky: 'exclude',
			inherit: false,
		},
	},
	scope: [ 'inserter' ],
},
```

Comment (apeatling) [reply packages/block-library/src/query/edit/index.js:71]: Thanks, this seems to be working correctly with the latest change:

Screenshot 2023-01-25 at 10 42 40 AM

**Comment by apeatling** [comment]
@ntsekouras How's this one looking now? I'd love to get this merged before the freeze. Thanks!

**Review by glendaviesnz — APPROVED** [review]
This tested well for me:

✅ Replace button showed dialog with search option
✅ Selecting pattern from dialog replaced query block in editor
✅ Unit tests passed

**Comment by apeatling** [comment]
I've fixed up the e2e tests for the query block pattern insertion to match the new list view. It's now passing. Waiting on this fix to [globally broken Gutenberg mobile tests](https://github.com/WordPress/gutenberg/pull/47465#issuecomment-1406840053), and will merge after assuming no further feedback.

**Comment by ntsekouras** [comment]
>@ntsekouras How's this one looking now? I'd love to get this merged before the freeze. Thanks!

Can you wait just a bit until Monday to test/review it? We need to make sure that the previous functionality with previews is preserved. Thank you!

File: packages/block-library/src/query/edit/pattern-selection-modal.js
Comment (ntsekouras) [inline packages/block-library/src/query/edit/pattern-selection-modal.js:77]: We could only wrap this component with `BlockContextProvider`.

**Review by ntsekouras — COMMENTED** [review]
Thanks @apeatling for your work and waiting me to review!

This looks good! I left only a minor comment and the only thing missing here is to style the `search` input to remain sticky, as we do in the Template Part relevant flows.

File: packages/block-library/src/query/edit/pattern-selection-modal.js
Comment (apeatling) [reply packages/block-library/src/query/edit/pattern-selection-modal.js:77]: Moved the location and retested the variations, all working great. 👍

**Comment by apeatling** [comment]
Thanks @ntsekouras, all fixed up, the search bar is now sticky and visually matches the template part block selection. 👍

**Comment by apeatling** [comment]
Going to merge this since it has approval and everything has been fixed up. :+1:

**Pull request merged**
